### PR TITLE
Use CodeMirror 6 in REPL

### PR DESCRIPTION
**Pull request by JLHwung**
In this PR we migrate the CodeMirror from 5 to 6 for REPL.

New features from this upgrade: 
- Find-and-replace support for the input source.
- Block level collapse-and-expand for both input source and output source

This is a vast rewrite of the `CodeMirror` bridging React component. Here are the UI tests I have performed manually:

- `preserveScrollPosition`: Input a 300-line text, scroll the output to the bottom, change the input, the output should preserve the scroll position
- `lineWrapping`: Input a 300-column text, toggle the "Line Wrap" option, the input should apply softwrap
- `readOnly`: The output editor should be readonly
- Dark theme: enable dark system theme, the REPL should provide a dark theme appearance; if system theme is switched to light, the REPL should provide a light theme.

Preview link: https://deploy-preview-3014--babel.netlify.app/repl

### CODE REVIEW

File: js/repl/CodeMirror.tsx [hunk 1]
```diff
@@ -1,138 +1,134 @@
...
-  mode: "text/jsx",
-  showCursorWhenSelecting: true,
-  styleActiveLine: true,
-  tabWidth: 2,
-  theme: preferDarkColorScheme() ? "darcula" : "default",
-};
+import React, { useRef, useEffect } from "react";
 
 type Props = {
-  autoFocus: boolean;
```
Comment (JLHwung) [inline js/repl/CodeMirror.tsx:20]: The `autoFocus` prop is removed because we are not using this prop.

File: js/repl/CodeMirror.tsx [hunk 2]
```diff
@@ -1,138 +1,134 @@
...
+      // });
+      if (options.lineWrapping) {
+        parentRef.current
+          .querySelector(".cm-content")
+          .classList.add("cm-lineWrapping");
+      } else {
+        parentRef.current
+          .querySelector(".cm-content")
+          .classList.remove("cm-lineWrapping");
+      }
```
Comment (JLHwung) [inline js/repl/CodeMirror.tsx:122]: I have to resort to DOM manipulations as the compartment update approach does not work, though it works for simple cases: https://codesandbox.io/p/sandbox/codemirror-6-demo-pl8dc

 Helps are welcomed.

File: js/repl/CodeMirror.tsx [hunk 3]
```diff
@@ -1,138 +1,134 @@
...
-    this._textAreaRef = ref;
-  };
+  return <div ref={parentRef} className="CodeMirror" />;
 }
 
 injectGlobal({
   ".CodeMirror": {
     height: "100% !important",
-    width: "100% !important",
-    WebkitOverflowScrolling: "touch",
```
Comment (JLHwung) [inline js/repl/CodeMirror.tsx:133]: This CSS property is long obsolete.

File: js/repl/CodeMirrorPanel.tsx [hunk 4]
```diff
@@ -30,7 +33,7 @@ export default function CodeMirrorPanel(props: Props) {
         <CodeMirror
           onChange={onChange}
           options={{
-            ...props.options,
```
Comment (JLHwung) [inline js/repl/CodeMirrorPanel.tsx:33]: The `fileSize` option is not used by the codemirror component. So here we explicitly specify the passing-through options.

**Review by liuxingbaoyu — APPROVED** [review]
Amazing!

**Comment by JLHwung** [comment]
Added a new polish commit: the dark theme will be automatically enabled whenever the system colour scheme is changed, and _without_ extra refresh.

**Pull request merged**